canals/pipeline/pipeline.py

```diff
@@ -420,57 +439,169 @@ def _check_max_loops(self, component_name: str):
-                # What if there are no components to wait for?
-                raise PipelineRuntimeError(
-                    f"'{name}' is stuck waiting for input, but there are no other components to run. "
-                    "This is likely a Canals bug. Open an issue at https://github.com/deepset-ai/canals."
-                )
+        ###############
+        # WAIT CHECKS #
+        ###############
+        if mandatory_input_sockets == received_input_sockets and skipped_optional_input_sockets.issubset(
+            optional_input_sockets
+        ):
+            # We received all of the inputs we need and some optional inputs have not been skipped
+            logger.debug(
+                "Component '%s' is waiting. All mandatory inputs received, some optional are not skipped: %s",
+                name,
+                optional_input_sockets - skipped_optional_input_sockets,
+            )
+            return "wait"
 
+        if any(self.graph.nodes[n]["visits"] == 0 for n in input_components.keys()):
+            # Some upstream component that must send input to the current component has yet to run.
+            # This happens when:
+            #   * Component has default values
+            #   * Component has some mandatory inputs
+            #   * Component has optional inputs that are not skipped
             logger.debug(
-                "Putting '%s' back in the queue, some inputs are missing (inputs to wait for: %s, inputs_received: %s)",
+                "Component '%s' is waiting. Missing inputs: %s",
                 name,
-                [f"{node}.{socket}" for node, socket in expected_inputs],
-                list(inputs.keys()),
+                set(input_components.values()),
             )
             return "wait"
 
+        ###############
+        # SKIP CHECKS #
+        ###############
+        if input_components and set(input_components.keys()).issubset(skipped_components):
+            # All mandatory inputs are skipped, skipping this too
+            logger.debug("Component '%s' is skipped. All upstream components are skipped.", name)
+            return "skip"
+
         logger.debug(
-            "Skipping '%s', upstream component didn't produce output "
-            "(upstream components: %s, expected inputs: %s, inputs received: %s)",
+            "Component '%s' is skipped. Can't determine whether it should run or wait. "
+            "Mandatory input sockets: %s, optional input sockets: %s, received input: %s, "
+            "input components: %s, skipped components: %s, skipped optional inputs: %s.",
             name,
-            list(nodes_to_wait_for),
-            [f"{node}.{socket}" for node, socket in expected_inputs],
+            mandatory_input_sockets,
+            optional_input_sockets,
             list(inputs.keys()),
+            list(input_components.keys()),
+            skipped_components,
+            skipped_optional_input_sockets,
         )
         return "skip"
```

Comment:
I'd be way more drastic: if you reach here the pipeline is on fire anyway, no point going further.

```suggestion
# Likely a bug
        logger.debug("""
            Component %s reached inconsistent state!
            Mandatory input sockets: %s
            Optional input sockets: %s
            Received input: %s
            Input components: %s
            Skipped components: %s
            Skipped optional inputs: %s""",
            name,
            mandatory_input_sockets,
            optional_input_sockets,
            list(inputs.keys()),
            list(input_components.keys()),
            skipped_components,
            skipped_optional_input_sockets,
        )
        raise PipelineRuntimeError("The component {name} reached an inconsistent state. This is likely a bug in Canals. Please open an issue here: https://github.com/deepset-ai/canals/issues/new, including the code that triggered this error and the DEBUG output of the run.")
```